### PR TITLE
Add an abstract environment 'defaults' in config for DRY

### LINKED ISSUE #249: config file does not follow DRY between environments
The config file currently contains a lot of duplication between the `test` and `development` environments. Suggest that we devise a mechanism to make an environment copy its configuration from another environment and then override particular settings.

### PR DESCRIPTION
**Pull request by bugos**
Fixes #249 and #289.

### CODE REVIEW

File: config/config.php [hunk 1]
```diff
@@ -27,36 +27,24 @@
             ],
             'forgot_password_exp_time' => 3600 * 24,
             'cli_max_width' => 120,
-            'email' => 'team@endofcodes.com'
+            'email' => 'team@endofcodes.com',
+            'root' => '/var/www/',
+            'base' => 'http://localhost/endofcodes/'
```
Comment (dionyziz) [inline config/config.php]: Why do `root` and `base` need to be specified since they can be deduced?

Comment (bugos) [inline config/config.php]: I think we should keep all used config keys in config.php. This way config.php can be used as a reference for our config system.

File: helpers/config.php [hunk 2]
```diff
@@ -9,21 +9,29 @@ function getBase() {
...
     }
+    function calculateConfigEntries( &$config ) {
+        $config[ 'root' ] = getcwd();
+        if ( isset( $_SERVER[ 'HTTP_HOST' ] ) ) {
+            $config[ 'base' ] = getBase();
+        }
+    }
+    function validateConfigEntries( $config ) {
+        if ( $config[ 'development' ][ 'db' ][ 'dbname' ] == $config[ 'test' ][ 'db' ][ 'dbname' ] ) {
+            throw new ModelValidationException( " Database name for development and testing is the same." );
```
Comment (dionyziz) [inline helpers/config.php:22]: nit: Remove space from the beginning of quotes

File: config/config.php (hunk 1 above)
Comment (dionyziz) [inline config/config.php]: Yes, but we want config-local to not contain a `root` and `base` key and for them to be derived automatically. But this way they will default to these values and they will need to be overwritten. I think these keys in particular should never be required to be specified in any config files – they are helpful to have for our application to function, but are not really part of the configuration formally.

File: helpers/config.php (hunk 2 above)
Comment (bugos) [inline helpers/config.php:22]: It is required for the error to be displayed properly.

Comment (bugos) [inline helpers/config.php:22]: btw config should be model

**Pull request merged**